From 46642cb36d63687fdd52b7161b192a3ac9e18393 Mon Sep 17 00:00:00 2001
From: Joan Torres Lopez <joantolo@redhat.com>
Date: Mon, 13 Apr 2026 17:06:05 +0200
Subject: [PATCH] manager: Schedule deferred plymouth quit on session
 registration

On headless or no-monitor systems, gnome-shell can register the session
but won't register the display, leaving plymouth running indefinitely.
That makes the system unusable because it didn't reach graphical.target.

Schedule a deferred plymouth quit when RegisterSession is called on a
local display, using the same REGISTER_DISPLAY_TIMEOUT (10s) as the
session wrappers. If RegisterDisplay arrives first, the timeout is
cancelled and plymouth is quit immediately with transition as before.

Part-of: <https://gitlab.gnome.org/GNOME/gdm/-/merge_requests/360>
---
 daemon/gdm-manager.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/daemon/gdm-manager.c b/daemon/gdm-manager.c
index 2f1b78a..b5685cd 100644
--- a/daemon/gdm-manager.c
+++ b/daemon/gdm-manager.c
@@ -94,6 +94,7 @@ struct GdmManagerPrivate
 
 #ifdef  WITH_PLYMOUTH
         guint                     plymouth_is_running : 1;
+        guint                     plymouth_quit_timeout_id;
 #endif
         guint                     did_automatic_login : 1;
 };
@@ -207,6 +208,21 @@ plymouth_quit_without_transition (void)
                            standard_error?: "");
         }
 }
+
+static gboolean
+plymouth_quit_timeout_cb (gpointer user_data)
+{
+        GdmManager *manager = user_data;
+
+        manager->priv->plymouth_quit_timeout_id = 0;
+
+        if (manager->priv->plymouth_is_running) {
+                plymouth_quit_without_transition ();
+                manager->priv->plymouth_is_running = FALSE;
+        }
+
+        return G_SOURCE_REMOVE;
+}
 #endif
 
 static char *
@@ -843,6 +859,23 @@ gdm_manager_handle_register_session (GdmDBusManager        *manager,
                       "session-registered", TRUE,
                       NULL);
 
+#ifdef WITH_PLYMOUTH
+        if (self->priv->plymouth_is_running) {
+                gboolean display_is_local = FALSE;
+
+                g_object_get (G_OBJECT (display),
+                              "is-local", &display_is_local,
+                              NULL);
+
+                if (display_is_local && self->priv->plymouth_quit_timeout_id == 0) {
+                        self->priv->plymouth_quit_timeout_id =
+                                g_timeout_add_seconds (REGISTER_DISPLAY_TIMEOUT,
+                                                       plymouth_quit_timeout_cb,
+                                                       self);
+                }
+        }
+#endif
+
         gdm_dbus_manager_complete_register_session (GDM_DBUS_MANAGER (manager),
                                                     invocation);
 
@@ -1647,6 +1680,7 @@ on_display_status_changed (GdmDisplay *display,
                         if (status == GDM_DISPLAY_MANAGED && quit_plymouth) {
                                 plymouth_quit_with_transition ();
                                 manager->priv->plymouth_is_running = FALSE;
+                                g_clear_handle_id (&manager->priv->plymouth_quit_timeout_id, g_source_remove);
                         }
 #endif
                         break;
@@ -1659,6 +1693,7 @@ on_display_status_changed (GdmDisplay *display,
                         if (quit_plymouth) {
                                 plymouth_quit_without_transition ();
                                 manager->priv->plymouth_is_running = FALSE;
+                                g_clear_handle_id (&manager->priv->plymouth_quit_timeout_id, g_source_remove);
                         }
 #endif
 
@@ -2620,6 +2655,10 @@ gdm_manager_stop (GdmManager *manager)
         }
 #endif
 
+#ifdef WITH_PLYMOUTH
+        g_clear_handle_id (&manager->priv->plymouth_quit_timeout_id, g_source_remove);
+#endif
+
         manager->priv->started = FALSE;
 }
 
@@ -2647,6 +2686,7 @@ gdm_manager_start (GdmManager *manager)
                 if (!manager->priv->show_local_greeter && manager->priv->plymouth_is_running) {
                         plymouth_quit_without_transition ();
                         manager->priv->plymouth_is_running = FALSE;
+                        g_clear_handle_id (&manager->priv->plymouth_quit_timeout_id, g_source_remove);
                 }
 #endif
                 if (manager->priv->xdmcp_factory != NULL) {
